File: code/go/2197.go
// https://leetcode.com/problems/replace-non-coprime-numbers-in-array/
func gcd(a, b int) int {
	for b != 0 {
		a, b = b, a%b
	}
	return a
}

func replaceNonCoprimes(nums []int) []int {
	stack := make([]int, 0, len(nums))
	for _, a := range nums {
		for a > 1 && len(stack) > 0 {
			n := len(stack) - 1
			b := stack[n]
			gcd_ := gcd(a, b)
			if gcd_ == 1 { break }
			a = a / gcd_ * b
			stack = stack[:n]
		}
		stack = append(stack, a)
	}
	return stack
}

File: import/new/2197.py
# https://leetcode.com/problems/replace-non-coprime-numbers-in-array/
class Solution:
    def replaceNonCoprimes(self, nums: List[int]) -> List[int]:
        stack = []
        gcd = math.gcd
        for a in nums:
            while a > 1 and stack:
                gcd_ = gcd(a, stack[-1])
                if gcd_ == 1:
                    break
                a *= stack.pop() // gcd_
            stack.append(a)
        return stack
